### ISSUE TITLE
[Feature request] Ability for LanguageTool to use external server

### ISSUE THREAD

**Issue by AimVaguely** [olivierkes/manuskript#1066]
Currently, the LanguageTool integration uses [language-check](https://pypi.org/project/language-check/) which the setup includes downloading LanguageTool itself and run a separate instance for that python instance (or something similar, I don't program python). 

This causes problems including:

- If you already have a running LanguageTool server running on your machine, then it downloads and runs two instances.
- You always need the java runtime accessible to Manuskript at all times, this means that the flatpak package doesn't work with LanguageTool.
- You need to run LanguageTool on the same machine you run Manuskript, which can cause problems because it seems that it uses a lot of memory.
- The version of LanguageTool that is being used is locked at what the package uses (which hasn't been updated in over a year).
- You can't use the [premium LanguageTool API that LibreOffice can](https://forum.languagetool.org/t/new-libreoffice-7-4-languagetool-remote-grammar-checker/8187?u=dnaber).

I was looking at the [repo for language-check](https://github.com/myint/language-check) and saw that it had [an issue about some of these problems](https://github.com/myint/language-check/issues/15) and there was [talk about how that package has been abandoned now](https://github.com/myint/language-check/issues/15#issuecomment-622234515), and how [there is a new version, which supports
local and remote servers](https://github.com/jxmorris12/language_tool_python#local-and-remote-servers). This seems not to have an option to use a premium API with the public instance, but that needs to be added to the library, not something that this application is responsible for.

So, this request includes two things:

- Migrate from the abandoned [language-check](https://pypi.org/project/language-check/) to [language-tool-python](https://pypi.org/project/language-tool-python/)
- Expose the local and remote servers settings to the user
- Package the [language-tool-python](https://pypi.org/project/language-tool-python/) package with Manuskript, not requiring it to be installed separately. (which is how the current one works)

Resolving these problems should also make LanguageTool with the Flatpak package.

**Comment by AimVaguely** [comment]
Thank you for the quick reply. 

Sorry about getting the information about what package/library is being used incorrect, I don't have much experience in python, so that was my mistake. Thank you for explaining the problem that the flatpak and snap are having, since I didn't have the full image.

You don't need to package the java runtime yourself, in flatpak you can add [SDK Extension for OpenJDK](https://github.com/flathub/org.freedesktop.Sdk.Extension.openjdk) as a dependency. This means multiple applications can share a single installation. After looking for a while, it [looks like snap can do a similar thing](https://snapcraft.io/blog/building-a-java-snap-by-example) (scroll down to `stage-packages` though the whole guide might be worth looking at). 

The even better solution would be for someone to package LanguageTool as an extension as well, but that might be a bit harder (I don't know). This would mean that flatpak can set up an instance of LanguageTool that can be shared by all applications, and it can deal with launching it in the background on boot. This might be what [this abandoned flatpak package](https://github.com/f0rki/org.languagetool.cli-flatpak) can do, I haven't tried it. Snap has [a package on LanguageTool that might provide it](https://snapcraft.io/languagetool). I'm not certain how these would integrate with the python package, I don't think they do at all (at least yet, they could use them though).

The easiest solution that would make it work would be exposing the setting to use an externally running instance, though (as you said). Though this would make the instructions to set it up more complicated, since the instructions are distro-specific, even with the flatpak version.

I agree with you that LanguageTool is too resource intensive to package as a default, but I, personally, used it before I switched to the flatpak version of Manuskript, and I miss it. I also already have a running LanguageTool server on my computer already (LibreOffice uses it), so it would take no extra system resources or disk space, for Manuskript to use it.

Sorry for the lengthy response, I just wanted to share my ideal solution. It would be amazing it the setting to use an external LanguageTool server was available, it would solve all the problems that I have (minus not having the ability to use the premium API, which the package/library doesn't support).

**Comment by TheJackiMonster** [comment]
> The even better solution would be for someone to package LanguageTool as an extension as well, but that might be a bit harder (I don't know). This would mean that flatpak can set up an instance of LanguageTool that can be shared by all applications, and it can deal with launching it in the background on boot. This might be what [this abandoned flatpak package](https://github.com/f0rki/org.languagetool.cli-flatpak) can do, I haven't tried it. Snap has [a package on LanguageTool that might provide it](https://snapcraft.io/languagetool). I'm not certain how these would integrate with the python package, I don't think they do at all (at least yet, they could use them though).

I agree that separating LanguageTool with Java would be the better solution. I assume the snap package is an application though. So I'm not sure if that can be utilized from another. For flatpak we would probably need some packaging which gets maintained.

Thank you for looking into the Java integration. I will try to look into it as well and check how much it would add to those packages.
